src/core/attribution/aggregation.js

```diff
@@ -25,11 +30,129 @@ export type ConnectionAggregation = {|
+    const aggregation: NodeAggregation = {
+      nodePrefix,
+      connections,
+      aggregationSummary: {size, aggregateScore},
+    };
+    result.push(aggregation);
+  }
+  if (connectionsFound !== xs.length) {
+    // It would be extremely misleading to silently return an aggregation that
+    // did not successfully aggregate every connection, as this would lead to
+    // pieces of the graph silently disappearing.
+    throw new Error(
+      "Invariant violation: connections found is not same as connections in"
+    );
+  }
+  return result;
+}
+
+/* The correctness of this function depends on the following conditions:
+ * 1. For every ScoredConnection in xs, there is exactly one NodePrefix for
+ * which the connection's source has that NodePrefix as a prefix.
+ * 2. For every ScoredConnection in xs that has an associated Edge, there is
+ * exactly one EdgePrefix for which the connection's edge's address has that
+ * EdgePrefix as a prefix.
+ * When the conditions are not met, the function will almost always error,
+ * but in certain cases it can return incorrect results.
+ */
 export function aggregateConnections(
   xs: $ReadOnlyArray<ScoredConnection>,
   edgePrefixes: $ReadOnlyArray<EdgeAddressT>,
   nodePrefixes: $ReadOnlyArray<NodeAddressT>
 ): $ReadOnlyArray<ConnectionAggregation> {
-  const _unused_stuff = [xs, edgePrefixes, nodePrefixes];
-  throw new Error("Not yet implemented");
+  let connectionsFound = 0;
+  const syntheticConnections = xs.filter(
+    (x) => x.connection.adjacency.type === "SYNTHETIC_LOOP"
+  );
+  const syntheticAggregation: ConnectionAggregation = {
+    connectionType: {type: "SYNTHETIC_LOOP"},
+    aggregationSummary: {
+      size: syntheticConnections.length,
+      aggregateScore: sumScoredConnections(syntheticConnections),
+    },
+    nodeAggregations: aggregateByNodePrefix(syntheticConnections, nodePrefixes),
+  };
+  connectionsFound += syntheticConnections.length;
+  function createAggregation(
+    connections: $ReadOnlyArray<ScoredConnection>,
+    type: "OUT_EDGE" | "IN_EDGE"
+  ) {
+    const matchingConnections = connections.filter(
+      (x) => x.connection.adjacency.type === type
+    );
+    const size = matchingConnections.length;
+    connectionsFound += size;
+    const aggregateScore = sumScoredConnections(matchingConnections);
+    const nodeAggregations = aggregateByNodePrefix(
+      matchingConnections,
+      nodePrefixes
+    );
+    return {
+      aggregationSummary: {size, aggregateScore},
+      nodeAggregations,
+    };
+  }
+  const result = [syntheticAggregation];
+  for (const edgePrefix of edgePrefixes) {
+    const withCorrectEdgePrefix = xs.filter(({connection: {adjacency}}) => {
+      if (adjacency.type === "IN_EDGE" || adjacency.type === "OUT_EDGE") {
+        return EdgeAddress.hasPrefix(adjacency.edge.address, edgePrefix);
+      }
+      return false;
+    });
+
+    // It would be much cleaner if createAggregation actually returned the
```

Comment:
You can work around the inference failure without too much difficulty:

```diff
diff --git a/src/core/attribution/aggregation.js b/src/core/attribution/aggregation.js
index 3307a6a..ca97b89 100644
--- a/src/core/attribution/aggregation.js
+++ b/src/core/attribution/aggregation.js
@@ -102,10 +102,12 @@ export function aggregateConnections(
   connectionsFound += syntheticConnections.length;
   function createAggregation(
     connections: $ReadOnlyArray<ScoredConnection>,
-    type: "OUT_EDGE" | "IN_EDGE"
-  ) {
+    connectionType:
+      | {|+type: "IN_EDGE", edgePrefix: EdgeAddressT|}
+      | {|+type: "OUT_EDGE", edgePrefix: EdgeAddressT|}
+  ): ConnectionAggregation {
     const matchingConnections = connections.filter(
-      (x) => x.connection.adjacency.type === type
+      (x) => x.connection.adjacency.type === connectionType.type
     );
     const size = matchingConnections.length;
     connectionsFound += size;
@@ -115,6 +117,7 @@ export function aggregateConnections(
       nodePrefixes
     );
     return {
+      connectionType,
       aggregationSummary: {size, aggregateScore},
       nodeAggregations,
     };
@@ -127,24 +130,12 @@ export function aggregateConnections(
       }
       return false;
     });
-
-    // It would be much cleaner if createAggregation actually returned the
-    // aggregation, but unfortunately a flow bug necessitates this more verbose
-    // construction.
-    const outCollation = createAggregation(withCorrectEdgePrefix, "OUT_EDGE");
-    const outAggregation: ConnectionAggregation = {
-      connectionType: {type: "OUT_EDGE", edgePrefix},
-      aggregationSummary: outCollation.aggregationSummary,
-      nodeAggregations: outCollation.nodeAggregations,
-    };
-    const inCollation = createAggregation(withCorrectEdgePrefix, "IN_EDGE");
-    const inAggregation: ConnectionAggregation = {
-      connectionType: {type: "IN_EDGE", edgePrefix},
-      aggregationSummary: inCollation.aggregationSummary,
-      nodeAggregations: inCollation.nodeAggregations,
-    };
-    result.push(outAggregation);
-    result.push(inAggregation);
+    result.push(
+      createAggregation(withCorrectEdgePrefix, {type: "OUT_EDGE", edgePrefix})
+    );
+    result.push(
+      createAggregation(withCorrectEdgePrefix, {type: "IN_EDGE", edgePrefix})
+    );
   }
   if (connectionsFound !== xs.length) {
     // It would be extremely misleading to silently return an aggregation that
```

Reply:
Fixed in the rewrite.